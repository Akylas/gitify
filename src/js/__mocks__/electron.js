window.Notification = function (title, options) {
  this.title = title;

  return {
    onclick: jest.fn()
  };
};

window.Audio = class Audio {
  constructor(path) {
    this.path = path;
  }

  play() {}
};

window.localStorage = {
  store: {},
  getItem: function (key) { return this.store[key]; },
  setItem: function (key, item) {
    this.store[key] = item;
  },
  removeItem: jest.fn()
};

window.alert = jest.fn();

const browserWindow = {
  loadURL: jest.fn(),
  webContents: {
    on: () => {},
  },
  on: () => {},
  close: jest.fn(),
  destroy: jest.fn()
};

module.exports = {
  remote: {
    BrowserWindow: () => browserWindow
  },
  ipcRenderer: {
    send: jest.fn()
  },
  shell: {
    openExternal: jest.fn()
  },
};
