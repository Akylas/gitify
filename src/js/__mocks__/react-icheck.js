import React from 'react';

export class Checkbox extends React.PureComponent {
  render() {
    return (
      <div>
        <strong>{this.props.label}</strong>
        <input
          type="checkbox"
          checked={this.props.defaultChecked}
          onChange={(evt) => this.props.onChange(evt)} />
      </div>
    );
  };
}

export class Radio extends React.PureComponent {
  render() {
    return (
      <div>
        <strong>{this.props.label}</strong>
        <input
          type="radio"
          value={this.props.value}
          checked={this.props.defaultChecked}
          onChange={(evt) => this.props.onChange(evt)} />
      </div>
    );
  };
}

export class RadioGroup extends React.PureComponent {
  render() {
    return (
      <div>
        {this.props.children}
      </div>
    );
  };
}
