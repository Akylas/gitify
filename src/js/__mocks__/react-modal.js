import React from 'react';

export default class Modal extends React.PureComponent {
  render() {
    return this.props.isOpen ? (
      <div>{this.props.children}</div>
    ) : null;
  };
}
