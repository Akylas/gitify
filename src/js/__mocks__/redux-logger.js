module.exports = function createLogger () {
  return ({ getState }) => (next) => (action) => {
    return next(action);
  };
};
