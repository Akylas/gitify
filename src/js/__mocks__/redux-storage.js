import { Map } from 'immutable';

export function createMiddleware(engine, actionBlacklist = [], actionWhitelist = []) {
  return ({ dispatch, getState }) => {
    return (next) => (action) => {
      return next(action);
    };
  };
};

export function createLoader() {
  const state = {
    auth: Map({
      token: 'LOGGED_IN_TOKEN'
    })
  };

  return () => new Promise(resolve => resolve(state));
}

export function reducer (clb) {
  return (state, action) => clb(state, action);
};
