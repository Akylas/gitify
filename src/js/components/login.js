const { ipcRenderer } = require('electron');

import React from 'react';
import { Redirect, Link } from 'react-router-dom';
import { connect } from 'react-redux';

import { authGithub, isUserEitherLoggedIn } from '../utils/helpers';

export class LoginPage extends React.Component {
  componentWillReceiveProps(nextProps) {
    if (nextProps.isEitherLoggedIn) {
      ipcRenderer.send('reopen-window');
    }
  }

  render() {
    if (this.props.isEitherLoggedIn) {
      return (
        <Redirect to="/" />
      );
    }

    return (
      <div className="container-fluid main-container login">
        <img className="img-responsive logo" src="images/gitify-logo-outline-dark.png" />
        <div className="desc">GitHub Notifications<br />in your menu bar.</div>
        <button
          className="btn btn-block btn-login"
          onClick={() => authGithub(undefined, this.props.dispatch)}
        >
          <i className="fa fa-github mr-2" /> Login to GitHub
        </button>

        <Link
          to="/enterpriselogin"
          className="btn btn-block btn-login mt-3">
          <i className="fa fa-github mr-2" /> Login to GitHub Enterprise
        </Link>
      </div>
    );
  }
};

export function mapStateToProps(state) {
  return {
    isEitherLoggedIn: isUserEitherLoggedIn(state.auth),
  };
};

export default connect(mapStateToProps)(LoginPage);
