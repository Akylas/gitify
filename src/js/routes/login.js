const { ipcRenderer } = require('electron');
import PropTypes from 'prop-types';

import React from 'react';
import { Redirect, Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faGithub } from '@fortawesome/free-brands-svg-icons';

import { authGithub, isUserEitherLoggedIn } from '../utils/helpers';
import LogoDark from '../components/logos/dark';

export class LoginPage extends React.Component {
  static propTypes = {
    isEitherLoggedIn: PropTypes.bool.isRequired,
    dispatch: PropTypes.func.isRequired,
  };

  componentWillReceiveProps(nextProps) {
    if (nextProps.isEitherLoggedIn) {
      ipcRenderer.send('reopen-window');
    }
  }

  render() {
    if (this.props.isEitherLoggedIn) {
      return <Redirect to="/" />;
    }

    return (
      <div className="container-fluid main-container login">
        <LogoDark className="mt-5" />

        <div className="desc">
          GitHub Notifications
          <br />
          in your menu bar.
        </div>
        <button
          className="btn btn-block btn-login"
          onClick={() => authGithub(undefined, this.props.dispatch)}
        >
          <FontAwesomeIcon className="mr-2" icon={faGithub} title="GitHub" />{' '}
          Login to GitHub
        </button>

        <Link to="/enterpriselogin" className="btn btn-block btn-login mt-3">
          <FontAwesomeIcon className="mr-2" icon={faGithub} title="GitHub" />{' '}
          Login to GitHub Enterprise
        </Link>
      </div>
    );
  }
}

export function mapStateToProps(state) {
  return {
    isEitherLoggedIn: isUserEitherLoggedIn(state.auth),
  };
}

export default connect(mapStateToProps)(LoginPage);
