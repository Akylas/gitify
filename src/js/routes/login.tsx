const { ipcRenderer } = require('electron');

import * as React from 'react';
import { Redirect, Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faGithub } from '@fortawesome/free-brands-svg-icons';

import { authGithub, isUserEitherLoggedIn } from '../utils/helpers';
import { LogoDark } from '../components/logos/dark';

interface IProps {
  isEitherLoggedIn: boolean;
  dispatch: any;
}

export class LoginPage extends React.Component<IProps> {
  constructor(props: IProps) {
    super(props);

    this.state = {
      isEitherLoggedIn: props.isEitherLoggedIn,
    };
  }

  static getDerivedStateFromProps(props, state) {
    if (props.isEitherLoggedIn) {
      ipcRenderer.send('reopen-window');
    }
    return {
      isEitherLoggedIn: props.isEitherLoggedIn,
    };
  }

  render() {
    if (this.props.isEitherLoggedIn) {
      return <Redirect to="/" />;
    }

    return (
      <div className="container-fluid main-container login">
        <LogoDark className="mt-5" />

        <div className="desc">
          GitHub Notifications
          <br />
          in your menu bar.
        </div>
        <button
          className="btn btn-block btn-login"
          onClick={() => authGithub(undefined, this.props.dispatch)}
        >
          <FontAwesomeIcon className="mr-2" icon={faGithub} title="GitHub" />{' '}
          Login to GitHub
        </button>

        <Link to="/enterpriselogin" className="btn btn-block btn-login mt-3">
          <FontAwesomeIcon className="mr-2" icon={faGithub} title="GitHub" />{' '}
          Login to GitHub Enterprise
        </Link>
      </div>
    );
  }
}

export function mapStateToProps(state) {
  return {
    isEitherLoggedIn: isUserEitherLoggedIn(state.auth),
  };
}

export default connect(mapStateToProps)(LoginPage);
