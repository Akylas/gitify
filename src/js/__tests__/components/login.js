import React from 'react'; // eslint-disable-line no-unused-vars
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import { Map } from 'immutable';
import { MemoryRouter } from 'react-router';

// import {authGithub} from '../../utils/Helpers';
import * as Helpers from '../../utils/Helpers';
import { LoginPage, mapStateToProps } from '../../components/login';
import { mockedEnterpriseAccounts } from '../../__mocks__/mockedData';

const { ipcRenderer, remote } = require('electron');
const BrowserWindow = remote.BrowserWindow;

describe('components/login.js', () => {
  const props = {
    dispatch: jest.fn()
  };

  beforeEach(function() {
    BrowserWindow().loadURL.mockReset();
    ipcRenderer.send.mockReset();
  });

  it('should test the mapStateToProps method', () => {
    const state = {
      auth: Map({
        token: '123456',
        failed: false,
        isFetching: false,
        enterpriseAccounts: mockedEnterpriseAccounts
      }),
    };

    const mappedProps = mapStateToProps(state);

    expect(mappedProps.isEitherLoggedIn).toBeTruthy();
  });

  it('should render itself & its children', () => {
    const props = {
      isEitherLoggedIn: false,
    };

    const tree = renderer.create(
      <MemoryRouter>
        <LoginPage {...props} />
      </MemoryRouter>
    );

    expect(tree).toMatchSnapshot();
  });

  it('should open the login window and get a code successfully (will-navigate)', () => {
    const code = '123123123';

    // console.log(Helpers);
    // spyOn(authGithub).and.callFake(() => {});
    // authGithub = jest.fn();
    spyOn(Helpers, 'authGithub');

    spyOn(BrowserWindow().webContents, 'on').and.callFake((event, callback) => {
      if (event === 'will-navigate') {
        callback('will-navigate', `http://www.github.com/?code=${code}`);
      }
    });

    const expectedUrl = 'https://github.com/login/oauth/' +
      'authorize?client_id=3fef4433a29c6ad8f22c&scope=user:email,notifications';

    const caseProps = {
      ...props,
      isEitherLoggedIn: false,
    };

    const wrapper = shallow(<LoginPage {...caseProps} />);

    expect(wrapper).toBeDefined();

    wrapper.find('button').simulate('click');

    // console.log(Helpers.authGithub.toString());

    expect(Helpers.authGithub).toHaveBeenCalledTimes(1);
    // expect(BrowserWindow().loadURL).toHaveBeenCalledTimes(1);
    // expect(BrowserWindow().loadURL).toHaveBeenCalledWith(expectedUrl);
    // expect(props.dispatch).toHaveBeenCalledTimes(1);
    // expect(props.dispatch.calls).toHaveBeenCalledWith(code);
  });

  it('should open the login window and get a code successfully (did-get-redirect-request)', () => {
    const code = '123123123';

    spyOn(BrowserWindow().webContents, 'on').and.callFake((event, callback) => {
      if (event === 'did-get-redirect-request') {
        callback('did-get-redirect-request', null, `http://www.github.com/?code=${code}`);
      }
    });

    const expectedUrl = 'https://github.com/login/oauth/' +
      'authorize?client_id=3fef4433a29c6ad8f22c&scope=user:email,notifications';

    const caseProps = {
      isEitherLoggedIn: false,
    };

    const wrapper = shallow(<LoginPage {...caseProps} />);

    expect(wrapper).toBeDefined();

    wrapper.find('.github').simulate('click');

    expect(BrowserWindow().loadURL).toHaveBeenCalledTimes(1);
    expect(BrowserWindow().loadURL).toHaveBeenCalledWith(expectedUrl);
    expect(props.loginUser).toHaveBeenCalledTimes(1);
    expect(props.loginUser).toHaveBeenCalledWith(code);
  });

  it('should open the login window and get an error', () => {
    const error = 'Oops! Something went wrong.';

    spyOn(BrowserWindow().webContents, 'on').and.callFake((event, callback) => {
      if (event === 'did-get-redirect-request') {
        callback('did-get-redirect-request', null, `http://www.github.com/?error=${error}`);
      }
    });

    const expectedUrl = 'https://github.com/login/oauth/' +
      'authorize?client_id=3fef4433a29c6ad8f22c&scope=user:email,notifications';

    const props = {
      loginUser: jest.fn(),
      token: null,
      response: {},
      failed: false,
      isFetching: false,
      settings: Map({
        baseUrl: 'github.com',
        clientId: '3fef4433a29c6ad8f22c',
        clientSecret: '9670de733096c15322183ff17ed0fc8704050379'
      }),
    };

    const wrapper = shallow(<LoginPage {...props} />);

    expect(wrapper).toBeDefined();

    wrapper.find('.github').simulate('click');

    expect(BrowserWindow().loadURL).toHaveBeenCalledTimes(1);
    expect(BrowserWindow().loadURL).toHaveBeenCalledWith(expectedUrl);
    expect(props.loginUser).not.toHaveBeenCalled();

    expect(alert).toHaveBeenCalledTimes(1);
    expect(alert).toHaveBeenCalledWith(
      'Oops! Something went wrong and we couldn\'t log you in using Github. Please try again.'
    );
  });

  it('should close the browser window before logging in', () => {
    spyOn(BrowserWindow(), 'on').and.callFake((event, callback) => {
      if (event === 'close') {
        callback();
      }
    });

    const props = {
      loginUser: jest.fn(),
      token: null,
      response: {},
      failed: false,
      isFetching: false,
      settings: Map({
        baseUrl: 'github.com',
        clientId: '3fef4433a29c6ad8f22c',
        clientSecret: '9670de733096c15322183ff17ed0fc8704050379'
      }),
    };

    const wrapper = shallow(<LoginPage {...props} />);

    expect(wrapper).toBeDefined();

    wrapper.find('.github').simulate('click');

    expect(BrowserWindow().loadURL).toHaveBeenCalledTimes(1);
    expect(props.loginUser).not.toHaveBeenCalled();
  });

  it('should redirect to notifications once logged in', () => {
    const props = {
      isLoggedIn: false,
      response: {},
      failed: false,
      isFetching: false,
      settings: Map({
        baseUrl: 'github.com',
        clientId: '3fef4433a29c6ad8f22c',
        clientSecret: '9670de733096c15322183ff17ed0fc8704050379'
      }),
    };

    const wrapper = shallow(<LoginPage {...props} />);

    expect(wrapper).toBeDefined();

    wrapper.setProps({
      isLoggedIn: true
    });

    expect(ipcRenderer.send).toHaveBeenCalledTimes(1);
    expect(ipcRenderer.send).toHaveBeenCalledWith('reopen-window');

    expect(wrapper.props().to).toEqual('/');
  });

  it('should request the github token if the oauth code is received', () => {
    const code = 'thisisacode';

    const props = {
      loginUser: jest.fn(),
      token: null,
      response: {},
      failed: false,
      isFetching: false
    };

    const wrapper = shallow(<LoginPage {...props} />);

    expect(wrapper).toBeDefined();

    wrapper.instance().requestGithubToken(code);
    expect(props.loginUser).toHaveBeenCalledTimes(1);
    expect(props.loginUser).toHaveBeenCalledWith(code);
  });
});
