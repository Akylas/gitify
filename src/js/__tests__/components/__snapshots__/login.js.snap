// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/login.js should render itself & its children 1`] = `
<div
  className="container-fluid main-container login"
>
  <img
    className="img-responsive logo"
    src="images/gitify-logo-outline-dark.png"
  />
  <div
    className="desc"
  >
    GitHub Notifications
    <br />
    in your menu bar.
  </div>
  <button
    className="btn btn-lg btn-block github"
    onClick={[Function]}
  >
    <i
      className="fa fa-github"
    />
     Log in to GitHub
  </button>
  <a
    className="btn btn-lg btn-block enterprise"
    href="/enterpriselogin"
    onClick={[Function]}
  >
    <i
      className="fa fa-github"
    />
     Enterprise
  </a>
</div>
`;
